In [1]:
from cmath import nan

import pandas as pd
import numpy as np



PERFORMANCE_COLUMNS = ['LOAN_SEQUENCE_NUMBER','MONTHLY_REPORTING_PERIOD','CURRENT_ACTUAL_UPB','CURRENT_LOAN_DELINQUENCY_STATUS','LOAN_AGE','REMAINING_MONTHS_TO_LEGAL_MATURITY','DEFECT_SETTLEMENT_DATE','MODIFICATION_FLAG','ZERO_BALANCE_CODE','ZERO_BALANCE_EFFECTIVE_DATE','CURRENT_INTEREST_RATE','CURRENT_NON_INTEREST_BEARING_UPB','DUE_DATE_OF_LAST_PAID_INSTALLMENT','MI_RECOVERIES','NET_SALE_PROCEEDS','NON_MI_RECOVERIES','TOTAL_EXPENSES','LEGAL_COSTS','MAINTENANCE_AND_PRESERVATION_COSTS','TAXES_AND_INSURANCE','MISCELLANEOUS_EXPENSES','ACTUAL_LOSS_CALCULATION','CUMULATIVE_MODIFICATION_COST','INTEREST_RATE_STEP_INDICATOR','PAYMENT_DEFERRAL_FLAG','ESTIMATED_LTV','ZERO_BALANCE_REMOVAL_UPB','DELINQUENT_ACCRUED_INTEREST','DELINQUENCY_DUE_TO_DISASTER','BORROWER_ASSISTANCE_STATUS_CODE','CURRENT_MONTH_MODIFICATION_COST','INTEREST_BEARING_UPB'
]


files = {
    "2007Q1": "/Users/macbookpro/platform/Backend/data/raw/historical_data_2007Q1/historical_data_time_2007Q1.txt",
    "2010Q1": "/Users/macbookpro/platform/Backend/data/raw/historical_data_2010Q1/historical_data_time_2010Q1.txt",
    "2020Q1": "/Users/macbookpro/platform/Backend/data/raw/historical_data_2020Q1/historical_data_time_2020Q1.txt",
    "2025Q1": "/Users/macbookpro/platform/Backend/data/raw/historical_data_2025/historical_data_2025Q1/historical_data_time_2025Q1.txt",

}

dfs = []
for vintage, path in files.items():
    df = pd.read_csv(
        path,
        sep="|",
        header=None,
        names=PERFORMANCE_COLUMNS,
        low_memory=False
    )
    df['VINTAGE'] = vintage
    dfs.append(df)
    print(f"{vintage} → {df.shape[0]:,} lignes / {df['LOAN_SEQUENCE_NUMBER'].nunique():,} prêts")

real = pd.concat(dfs, ignore_index=True)


2007Q1 → 19,386,281 lignes / 302,331 prêts
2010Q1 → 23,179,374 lignes / 360,856 prêts
2020Q1 → 20,840,433 lignes / 517,072 prêts
2025Q1 → 1,427,609 lignes / 210,749 prêts


# Restauration des Headers

In [2]:
real.head()

,LOAN_SEQUENCE_NUMBER,MONTHLY_REPORTING_PERIOD,CURRENT_ACTUAL_UPB,CURRENT_LOAN_DELINQUENCY_STATUS,LOAN_AGE,REMAINING_MONTHS_TO_LEGAL_MATURITY,DEFECT_SETTLEMENT_DATE,MODIFICATION_FLAG,ZERO_BALANCE_CODE,ZERO_BALANCE_EFFECTIVE_DATE,...,INTEREST_RATE_STEP_INDICATOR,PAYMENT_DEFERRAL_FLAG,ESTIMATED_LTV,ZERO_BALANCE_REMOVAL_UPB,DELINQUENT_ACCRUED_INTEREST,DELINQUENCY_DUE_TO_DISASTER,BORROWER_ASSISTANCE_STATUS_CODE,CURRENT_MONTH_MODIFICATION_COST,INTEREST_BEARING_UPB,VINTAGE
0,F07Q10000001,200704,169000.0,0,0,240.0,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,169000.0,2007Q1
1,F07Q10000001,200705,168000.0,0,1,239.0,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,168000.0,2007Q1
2,F07Q10000001,200706,168000.0,0,2,238.0,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,168000.0,2007Q1
3,F07Q10000001,200707,168000.0,0,3,237.0,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,168000.0,2007Q1
4,F07Q10000001,200708,167000.0,0,4,236.0,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,167000.0,2007Q1


## Remplacer les valeur de convention par les NaN

- LOAN SEQUENCE NUMBER : identifiant unique du pret, documentation n'assume aucune valeur de convention

- MONTHLY_REPORTING_PERIOD : mois d'observation, documentation n'assume aucune valeur de convention, mais transformation en datetime

In [3]:
real.MONTHLY_REPORTING_PERIOD = pd.to_datetime(real.MONTHLY_REPORTING_PERIOD, format='%Y%m')

- CURRENT ACTUAL UPB : on assume aucune valeur de convention

- CURRENT_LOAN_DELINQUENCY_STATUS : on assume aucune valeur de convention

- LOAN_AGE : on assume aucune valeur de convention

In [4]:
real.LOAN_AGE.unique()

array([  0,   1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,
        13,  14,  15,  16,  17,  18,  19,  20,  21,  22,  23,  24,  25,
        26,  27,  28,  29,  30,  31,  32,  33,  34,  35,  36,  37,  38,
        39,  40,  41,  42,  43,  44,  45,  46,  47,  48,  49,  50,  51,
        52,  53,  54,  55,  56,  57,  58,  59,  60,  61,  62,  63,  64,
        65,  66,  67,  68,  69,  70,  71,  72,  73,  74,  75,  76,  77,
        78,  79,  80,  81,  82,  83,  84,  85,  86,  87,  88,  89,  90,
        91,  92,  93,  94,  95,  96,  97,  98,  99, 100, 101, 102, 103,
       104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116,
       117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129,
       130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142,
       143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155,
       156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168,
       169, 170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 18

- REMAINING_MONTHS_TO_LEGAL_MATURITY : on assume aucune valeur de convention

- DEFECT_SETTLEMENT_DATE : on assume aucune valeur de convention

In [5]:
real.DEFECT_SETTLEMENT_DATE.unique()

array([    nan, 201008., 200709., 200706., 201001., 200809., 201403.,
       201302., 200805., 201111., 201306., 201207., 201408., 201212.,
       200911., 201012., 201109., 201003., 200806., 201312., 201105.,
       201311., 201206., 200811., 200907., 201211., 201304., 201210.,
       201310., 201006., 200705., 201101., 201502., 200912., 200904.,
       201011., 201301., 201305., 201402., 201409., 201404., 200803.,
       201203., 201107., 201009., 201208., 200712., 201005., 201401.,
       201112., 201202., 201104., 201303., 201110., 201209., 201504.,
       201307., 201108., 201308., 200906., 200903., 200905., 201509.,
       201106., 201103., 200812., 201410., 200707., 201510., 200901.,
       200902., 200810., 201002., 201205., 201007., 201102., 202506.,
       201204., 201010., 200908., 201004., 201309., 200808., 200910.,
       200708., 200711., 202205., 201406., 201812., 200703., 201506.,
       201902., 201507., 200909., 202003., 200710., 201612., 201201.,
       201702., 2008

- MODIFICATION_FLAG : on assume aucune valeur de convention, la colonne est a supprimer
- regle : [Y,P,N]


In [6]:
real['MODIFICATION_FLAG'] = real['MODIFICATION_FLAG'].astype(object)
real.MODIFICATION_FLAG.unique()

array([nan, 'Y', 'P'], dtype=object)

- ZERO_BALANCE_CODE :

        01 = Prepaid or Matured (Voluntary Payoff)
        • 02 = Third Party Sale
        • 03 = Short Sale or
         Charge Off
        • 09 = REO Disposition
        • 15 = Whole Loan sales
        • 16 = Reperforming loan
        securitizations
        • 96 = Defect prior to other termination event
    On verifie egalement la coherence avec Les date

In [7]:
real['ZERO_BALANCE_CODE'] = real['ZERO_BALANCE_CODE'].replace(np.nan, 0).astype(object)
real.ZERO_BALANCE_CODE.unique()

array([0.0, 1.0, 2.0, 9.0, 16.0, 96.0, 3.0, 15.0], dtype=object)

- ZERO_BALANCE_EFFECTIVE_DATE : on assume aucune valeur de convention, ls valeur sont des mois, transformation en datetime

In [8]:
real.ZERO_BALANCE_EFFECTIVE_DATE = pd.to_datetime(real.ZERO_BALANCE_EFFECTIVE_DATE, format='%Y%m')
real.ZERO_BALANCE_EFFECTIVE_DATE.unique()

<DatetimeArray>
[                'NaT', '2008-04-01 00:00:00', '2022-02-01 00:00:00',
 '2020-12-01 00:00:00', '2016-08-01 00:00:00', '2017-06-01 00:00:00',
 '2010-09-01 00:00:00', '2009-01-01 00:00:00', '2009-03-01 00:00:00',
 '2009-07-01 00:00:00',
 ...
 '2023-12-01 00:00:00', '2025-09-01 00:00:00', '2025-03-01 00:00:00',
 '2024-01-01 00:00:00', '2023-09-01 00:00:00', '2025-02-01 00:00:00',
 '2025-08-01 00:00:00', '2023-06-01 00:00:00', '2024-05-01 00:00:00',
 '2025-10-01 00:00:00']
Length: 227, dtype: datetime64[ns]

- CURRENT_INTEREST_RATE : RAS

In [9]:
real.CURRENT_INTEREST_RATE.unique()

array([7.375, 6.125, 7.125, ..., 7.695, 7.804, 8.55 ], shape=(4117,))

- CURRENT_NON_INTEREST_BEARING_UPB : RAS

In [10]:
real.CURRENT_NON_INTEREST_BEARING_UPB.unique()

array([    0.  ,  7600.  , 40700.  , ...,  9328.58,  4583.45,  7812.33],
      shape=(24943,))

- DUE_DATE_OF_LAST_PAID_INSTALLMENT : RAS, transformation en datetime

In [11]:
real.DUE_DATE_OF_LAST_PAID_INSTALLMENT = pd.to_datetime(real.DUE_DATE_OF_LAST_PAID_INSTALLMENT, format='%Y%m')
real.DUE_DATE_OF_LAST_PAID_INSTALLMENT.unique()

<DatetimeArray>
[                'NaT', '2022-01-01 00:00:00', '2020-11-01 00:00:00',
 '2010-07-01 00:00:00', '2024-09-01 00:00:00', '2021-10-01 00:00:00',
 '2021-01-01 00:00:00', '2008-12-01 00:00:00', '2015-04-01 00:00:00',
 '2021-05-01 00:00:00',
 ...
 '2050-04-01 00:00:00', '2040-04-01 00:00:00', '2027-01-01 00:00:00',
 '2035-04-01 00:00:00', '2035-09-01 00:00:00', '2050-05-01 00:00:00',
 '2026-07-01 00:00:00', '2027-06-01 00:00:00', '2055-02-01 00:00:00',
 '2055-03-01 00:00:00']
Length: 258, dtype: datetime64[ns]

- MI_RECOVERIES : Que des Nan, #todo trouver son mode d'obtention

In [12]:
real.MI_RECOVERIES.unique()

array([     nan,     0.  , 52009.18, ...,  7471.06, 16172.78, 17274.53],
      shape=(5657,))

- NET_SALE_PROCEEDS : Que des Nan, #todo trouver son mode d'obtention

In [13]:
real.NET_SALE_PROCEEDS.isna().mean()

np.float64(0.9995512672985469)

In [14]:
real.NET_SALE_PROCEEDS.unique()

array([      nan, 209033.75,  28601.63, ..., 316504.38, 112997.98,
        74548.41], shape=(27705,))

- NON_MI_RECOVERIE : que des Nan, #todo trouver son mode d'obtention

In [15]:
real.NON_MI_RECOVERIES.unique()

array([    nan, 1640.39,  109.99, ..., 4868.52,  733.64, 6600.78],
      shape=(23261,))

- TOTAL_EXPENSES : Que des NAN

In [16]:
real.TOTAL_EXPENSES.unique()

array([      nan, -17328.69, -16238.77, ..., -38832.8 , -18869.53,
        -2124.39], shape=(27906,))

- Legal Costs : Que des Nan

In [17]:
real.LEGAL_COSTS.unique()

array([      nan,  -2506.11,  -3735.  , ...,  -5948.36, -12608.76,
        -6191.59], shape=(23674,))

- Maintenance and Preservation Costs : Que des Nan

In [18]:
real.MAINTENANCE_AND_PRESERVATION_COSTS.unique()

array([      nan,  -2000.  ,  -9750.62, ...,  -4982.01, -42870.6 ,
         -804.95], shape=(18167,))

- Tax and Insurance : Que des Nan

In [19]:
real.TAXES_AND_INSURANCE.unique()

array([      nan, -12822.58,  -1708.  , ..., -16202.79, -23055.81,
       -12398.94], shape=(26051,))

- Miscellaneous Expenses : Que des Nan

In [20]:
real.MISCELLANEOUS_EXPENSES.unique()

array([     nan,     0.  , -1045.15, ...,  -770.71, -2363.28, -2124.39],
      shape=(7242,))

- Actual Loss Calculation : Que des Nan

In [21]:
real.ACTUAL_LOSS_CALCULATION.unique()

array([       nan, -133189.29,  -29735.24, ...,  -52119.86,  -17432.71,
         -2124.39], shape=(28907,))

- Cumulative Modification Cost : aucune convention signaler

In [22]:
real.CUMULATIVE_MODIFICATION_COST.unique()

array([         nan, 9.160289e+04, 6.973607e+04, ..., 4.206000e+01,
       2.027600e+02, 2.779000e+01], shape=(34887,))

- Interest Rate Step Indicator : aucune convention signaler

In [23]:
real.INTEREST_RATE_STEP_INDICATOR.unique()

array([nan, 'N', 'Y'], dtype=object)

- Payment deferment flag : aucune convention signaler

In [24]:
real.PAYMENT_DEFERRAL_FLAG.unique()

array([nan, 'Y', 'P'], dtype=object)

- Estimated LTV: convention detecter

In [25]:
real.ESTIMATED_LTV.isna().mean()

np.float64(0.550467128228088)

In [26]:
real['ESTIMATED_LTV'] = real['ESTIMATED_LTV'].replace(999, np.nan)

In [27]:
real.ESTIMATED_LTV.isna().mean()

np.float64(0.5869894477866965)

- Zero Balance Removal UPB : aucune convention signaler



In [28]:
real.ZERO_BALANCE_REMOVAL_UPB.isna().mean()

np.float64(0.9855969342608983)

- Delinquent Accrued Interest : aucune convention signaler


In [29]:
real.DELINQUENT_ACCRUED_INTEREST.unique()

array([     nan, 38981.44,  5844.01, ..., 40900.92,  5496.22,  2700.29],
      shape=(28779,))

- Borrower BORROWER ASSISTANCE STATUS CODE

In [30]:
real['BORROWER_ASSISTANCE_STATUS_CODE'] = real['BORROWER_ASSISTANCE_STATUS_CODE'].replace(np.nan,'N').astype(object)

- CURRENT MONTH MODIFICATION COST -

In [31]:
real.CURRENT_MONTH_MODIFICATION_COST.unique()

array([    nan,  579.98,  574.91, ..., -112.88,  -44.92,  818.31],
      shape=(151334,))

- INTEREST BEARING UPB

In [32]:
real.INTEREST_BEARING_UPB.unique()

array([169000.  , 168000.  , 167000.  , ..., 762073.91, 326866.01,
       120506.75], shape=(24453185,))

Delinquency due to deasester

In [33]:
real['DELINQUENCY_DUE_TO_DISASTER'] = real['DELINQUENCY_DUE_TO_DISASTER'].replace(np.nan,'N').astype(object)

# Cleaning

-       Suppresion des variable Post-decision
-       les variable a 100 % de NaN seront supprimer si il s'agit de variable Post-decision
-       Imputation des variables restantes par mediane mode

In [34]:
real.isna().mean()

LOAN_SEQUENCE_NUMBER                  0.000000
MONTHLY_REPORTING_PERIOD              0.000000
CURRENT_ACTUAL_UPB                    0.000000
CURRENT_LOAN_DELINQUENCY_STATUS       0.000000
LOAN_AGE                              0.000000
REMAINING_MONTHS_TO_LEGAL_MATURITY    0.000006
DEFECT_SETTLEMENT_DATE                0.999880
MODIFICATION_FLAG                     0.975416
ZERO_BALANCE_CODE                     0.000000
ZERO_BALANCE_EFFECTIVE_DATE           0.985597
CURRENT_INTEREST_RATE                 0.000000
CURRENT_NON_INTEREST_BEARING_UPB      0.000000
DUE_DATE_OF_LAST_PAID_INSTALLMENT     0.993739
MI_RECOVERIES                         0.999552
NET_SALE_PROCEEDS                     0.999551
NON_MI_RECOVERIES                     0.999552
TOTAL_EXPENSES                        0.999552
LEGAL_COSTS                           0.999552
MAINTENANCE_AND_PRESERVATION_COSTS    0.999552
TAXES_AND_INSURANCE                   0.999552
MISCELLANEOUS_EXPENSES                0.999552
ACTUAL_LOSS_C

In [35]:
ELIGIBLE_COLUMNS = [
    "LOAN_SEQUENCE_NUMBER",
    "MONTHLY_REPORTING_PERIOD",
    "CURRENT_ACTUAL_UPB",
    "CURRENT_LOAN_DELINQUENCY_STATUS",
    "LOAN_AGE",
    "REMAINING_MONTHS_TO_LEGAL_MATURITY",
    "MODIFICATION_FLAG",
    "ZERO_BALANCE_CODE",
    "ZERO_BALANCE_EFFECTIVE_DATE",
    "CURRENT_INTEREST_RATE",
    "CURRENT_NON_INTEREST_BEARING_UPB",
    "DUE_DATE_OF_LAST_PAID_INSTALLMENT",
    "INTEREST_RATE_STEP_INDICATOR",
    "ESTIMATED_LTV",
    "DELINQUENCY_DUE_TO_DISASTER",
    "BORROWER_ASSISTANCE_STATUS_CODE",
    "INTEREST_BEARING_UPB",
]

In [36]:
ELIGIBLE_COLUMNS = ELIGIBLE_COLUMNS + ['VINTAGE','LOAN_SEQUENCE_NUMBER']
to_drop = [col for col in real.columns if   col not in ELIGIBLE_COLUMNS]
to_drop

['DEFECT_SETTLEMENT_DATE',
 'MI_RECOVERIES',
 'NET_SALE_PROCEEDS',
 'NON_MI_RECOVERIES',
 'TOTAL_EXPENSES',
 'LEGAL_COSTS',
 'MAINTENANCE_AND_PRESERVATION_COSTS',
 'TAXES_AND_INSURANCE',
 'MISCELLANEOUS_EXPENSES',
 'ACTUAL_LOSS_CALCULATION',
 'CUMULATIVE_MODIFICATION_COST',
 'PAYMENT_DEFERRAL_FLAG',
 'ZERO_BALANCE_REMOVAL_UPB',
 'DELINQUENT_ACCRUED_INTEREST',
 'CURRENT_MONTH_MODIFICATION_COST']

- correction des Variable Post decision

In [37]:
real['DEFECT_SETTLEMENT_DATE'] = pd.to_datetime(real['DEFECT_SETTLEMENT_DATE'], format='%Y%m')
real['NET_SALE_PROCEEDS'] = real['NET_SALE_PROCEEDS'].astype(object).replace(np.nan,'U')
real['PAYMENT_DEFERRAL_FLAG'] = real['PAYMENT_DEFERRAL_FLAG'].astype(object).replace(np.nan,'N')


- gestion des valeur manquante

In [38]:
real[to_drop].isna().mean()

DEFECT_SETTLEMENT_DATE                0.999880
MI_RECOVERIES                         0.999552
NET_SALE_PROCEEDS                     0.000000
NON_MI_RECOVERIES                     0.999552
TOTAL_EXPENSES                        0.999552
LEGAL_COSTS                           0.999552
MAINTENANCE_AND_PRESERVATION_COSTS    0.999552
TAXES_AND_INSURANCE                   0.999552
MISCELLANEOUS_EXPENSES                0.999552
ACTUAL_LOSS_CALCULATION               0.999551
CUMULATIVE_MODIFICATION_COST          0.999439
PAYMENT_DEFERRAL_FLAG                 0.000000
ZERO_BALANCE_REMOVAL_UPB              0.985597
DELINQUENT_ACCRUED_INTEREST           0.999551
CURRENT_MONTH_MODIFICATION_COST       0.966063
dtype: float64

In [39]:
# Ces variables sont-elles renseignées uniquement sur les prêts clôturés ?
mask_cloture = real['ZERO_BALANCE_CODE'].isin([2, 3, 9])

print(real[mask_cloture][['MI_RECOVERIES', 'NON_MI_RECOVERIES',
                           'TOTAL_EXPENSES', 'ACTUAL_LOSS_CALCULATION',
                           'DELINQUENT_ACCRUED_INTEREST']].isna().mean().round(3))

MI_RECOVERIES                  0.072
NON_MI_RECOVERIES              0.072
TOTAL_EXPENSES                 0.072
ACTUAL_LOSS_CALCULATION        0.070
DELINQUENT_ACCRUED_INTEREST    0.070
dtype: float64


In [40]:
# Voir ces cas
mask_cloture = real['ZERO_BALANCE_CODE'].isin([2, 3, 9])
mask_nan = real['ACTUAL_LOSS_CALCULATION'].isna()

print(real[mask_cloture & mask_nan][['LOAN_SEQUENCE_NUMBER',
                                      'ZERO_BALANCE_CODE',
                                      'ZERO_BALANCE_EFFECTIVE_DATE',
                                      'ACTUAL_LOSS_CALCULATION']].head(10))

       LOAN_SEQUENCE_NUMBER ZERO_BALANCE_CODE ZERO_BALANCE_EFFECTIVE_DATE  \
51150          F07Q10000873               9.0                  2014-01-01   
57601          F07Q10001015               3.0                  2011-08-01   
74877          F07Q10001502               9.0                  2012-12-01   
113121         F07Q10002358               3.0                  2010-09-01   
121186         F07Q10002575               9.0                  2011-08-01   
133595         F07Q10002832               3.0                  2011-11-01   
172180         F07Q10003634               9.0                  2010-12-01   
182439         F07Q10003905               9.0                  2013-09-01   
191479         F07Q10004096               9.0                  2008-01-01   
192616         F07Q10004120               9.0                  2009-07-01   

        ACTUAL_LOSS_CALCULATION  
51150                       NaN  
57601                       NaN  
74877                       NaN  
113121          

In [41]:
# Variables de recouvrement — imputer à 0 sur les prêts clôturés en défaut
cols_recouvrement = [
    'MI_RECOVERIES', 'NON_MI_RECOVERIES', 'TOTAL_EXPENSES',
    'LEGAL_COSTS', 'MAINTENANCE_AND_PRESERVATION_COSTS',
    'TAXES_AND_INSURANCE', 'MISCELLANEOUS_EXPENSES',
    'ACTUAL_LOSS_CALCULATION', 'DELINQUENT_ACCRUED_INTEREST'
]

mask_cloture = real['ZERO_BALANCE_CODE'].isin([2, 3, 9])

for col in cols_recouvrement:
    real.loc[mask_cloture & real[col].isna(), col] = 0

# Vérification
real[mask_cloture][cols_recouvrement].isna().mean().round(3)

MI_RECOVERIES                         0.0
NON_MI_RECOVERIES                     0.0
TOTAL_EXPENSES                        0.0
LEGAL_COSTS                           0.0
MAINTENANCE_AND_PRESERVATION_COSTS    0.0
TAXES_AND_INSURANCE                   0.0
MISCELLANEOUS_EXPENSES                0.0
ACTUAL_LOSS_CALCULATION               0.0
DELINQUENT_ACCRUED_INTEREST           0.0
dtype: float64

In [42]:
# CUMULATIVE_MODIFICATION_COST → coût cumulé des modifications. NaN = jamais modifié.
real['CUMULATIVE_MODIFICATION_COST'] = real['CUMULATIVE_MODIFICATION_COST'].fillna(0)

#CURRENT_MONTH_MODIFICATION_COST → coût de modification ce mois. NaN = pas de modification ce mois.
real['CURRENT_MONTH_MODIFICATION_COST'] = real['CURRENT_MONTH_MODIFICATION_COST'].fillna(0)



real['CUMULATIVE_MODIFICATION_COST'] = real['CUMULATIVE_MODIFICATION_COST'].fillna(0)
real['CURRENT_MONTH_MODIFICATION_COST'] = real['CURRENT_MONTH_MODIFICATION_COST'].fillna(0)

print(real[['CUMULATIVE_MODIFICATION_COST',
            'CURRENT_MONTH_MODIFICATION_COST']].isna().sum())

CUMULATIVE_MODIFICATION_COST       0
CURRENT_MONTH_MODIFICATION_COST    0
dtype: int64



        Delinquent Accrued Interest = (Zero Balance Removal UPB – Non Interest bearing UPB) * Min(Current Interest rate – 0.35, Current Interest
        Rate – Servicing Fee) * ( Months between Last Principal & Interest paid-to date and zero balance date ) * 30/360/100.

In [43]:
#real = real.drop(to_drop, axis=1)

In [44]:
real.isna().mean()

LOAN_SEQUENCE_NUMBER                  0.000000
MONTHLY_REPORTING_PERIOD              0.000000
CURRENT_ACTUAL_UPB                    0.000000
CURRENT_LOAN_DELINQUENCY_STATUS       0.000000
LOAN_AGE                              0.000000
REMAINING_MONTHS_TO_LEGAL_MATURITY    0.000006
DEFECT_SETTLEMENT_DATE                0.999880
MODIFICATION_FLAG                     0.975416
ZERO_BALANCE_CODE                     0.000000
ZERO_BALANCE_EFFECTIVE_DATE           0.985597
CURRENT_INTEREST_RATE                 0.000000
CURRENT_NON_INTEREST_BEARING_UPB      0.000000
DUE_DATE_OF_LAST_PAID_INSTALLMENT     0.993739
MI_RECOVERIES                         0.999520
NET_SALE_PROCEEDS                     0.000000
NON_MI_RECOVERIES                     0.999520
TOTAL_EXPENSES                        0.999520
LEGAL_COSTS                           0.999520
MAINTENANCE_AND_PRESERVATION_COSTS    0.999520
TAXES_AND_INSURANCE                   0.999520
MISCELLANEOUS_EXPENSES                0.999520
ACTUAL_LOSS_C

Imputation des nan pour les Varibale eligible

In [45]:
orig = pd.read_csv('/Users/macbookpro/platform/Backend/data/working/origination_v01.csv')

/var/folders/gy/2mbpv0kx5jz22fry28cqlnkc0000gn/T/ipykernel_26555/3832005183.py:1: DtypeWarning: Columns (24,25,29) have mixed types. Specify dtype option on import or set low_memory=False.
  orig = pd.read_csv('/Users/macbookpro/platform/Backend/data/working/origination_v01.csv')


REMAINING_MONTHS_TO_LEGAL_MATURITY

        IL manque des informations concernnt le nombre de mois restant avant la date de clôture
        pour le corriger on recuper les Date de Maturite presente a l'origination et recalcule



In [46]:
# REMAINING = ORIGINAL_LOAN_TERM - LOAN_AGE
# mais nécessite la jointure pour ORIGINAL_LOAN_TERM
print(real['REMAINING_MONTHS_TO_LEGAL_MATURITY'].isna().sum())
# Sans jointure : vérifier si on peut déduire depuis les autres lignes du même prêt
mask = real['REMAINING_MONTHS_TO_LEGAL_MATURITY'].isna()
real[mask][['LOAN_SEQUENCE_NUMBER', 'LOAN_AGE', 'REMAINING_MONTHS_TO_LEGAL_MATURITY']].head(10)

381


,LOAN_SEQUENCE_NUMBER,LOAN_AGE,REMAINING_MONTHS_TO_LEGAL_MATURITY
1719853,F07Q10038033,0,NaN
1719854,F07Q10038033,1,NaN
1719855,F07Q10038033,2,NaN
1719856,F07Q10038033,3,NaN
1719857,F07Q10038033,4,NaN
1719858,F07Q10038033,5,NaN
1719859,F07Q10038033,6,NaN
1719860,F07Q10038033,7,NaN
1719861,F07Q10038033,8,NaN
1719862,F07Q10038033,9,NaN


In [47]:
# Cibler uniquement le prêt concerné
target_loan = real[real['REMAINING_MONTHS_TO_LEGAL_MATURITY'].isna()]['LOAN_SEQUENCE_NUMBER'].unique()

orig_term = orig[orig['LOAN_SEQUENCE_NUMBER'].isin(target_loan)][['LOAN_SEQUENCE_NUMBER', 'ORIGINAL_LOAN_TERM']]

# Jointure ciblée
real = real.merge(orig_term, on='LOAN_SEQUENCE_NUMBER', how='left')

# Imputation
mask = real['REMAINING_MONTHS_TO_LEGAL_MATURITY'].isna()
real.loc[mask, 'REMAINING_MONTHS_TO_LEGAL_MATURITY'] = \
    real.loc[mask, 'ORIGINAL_LOAN_TERM'] - real.loc[mask, 'LOAN_AGE']

# Nettoyage
real = real.drop(columns=['ORIGINAL_LOAN_TERM'])

real['REMAINING_MONTHS_TO_LEGAL_MATURITY'].isna().sum()


np.int64(0)

LTV

        ESTIMATED_LTV = CURRENT_ACTUAL_UPB / valeur du bien
        La valeur du bien n'est pas directement disponible, mais on peut la déduire depuis l'origination

        valeur du bien à l'origination = ORIGINAL_UPB / (LTV / 100)

        ESTIMATED_LTV = CURRENT_ACTUAL_UPB / (ORIGINAL_UPB / (LTV / 100))

        C'est un proxy — on suppose que la valeur du bien reste celle de l'origination, ce qui est une approximation conservatrice.
        On a besoin de ORIGINAL_UPB et LTV depuis l'origination

In [48]:
# Vérifier la distribution des NaN par millésime
real.groupby('VINTAGE')['ESTIMATED_LTV'].apply(lambda x: x.isna().mean()).round(3)

VINTAGE
2007Q1    0.926
2010Q1    0.804
2020Q1    0.063
2025Q1    0.115
Name: ESTIMATED_LTV, dtype: float64

In [49]:
# Cibler uniquement les prêts avec ESTIMATED_LTV manquant
target_loans = real[real['ESTIMATED_LTV'].isna()]['LOAN_SEQUENCE_NUMBER'].unique()

# Extraire depuis l'origination
orig_ltv = orig[orig['LOAN_SEQUENCE_NUMBER'].isin(target_loans)][
    ['LOAN_SEQUENCE_NUMBER', 'ORIGINAL_UPB', 'LTV']
]

# Jointure ciblée
real = real.merge(orig_ltv, on='LOAN_SEQUENCE_NUMBER', how='left')

# Calcul du proxy
mask = real['ESTIMATED_LTV'].isna()
valeur_bien = real.loc[mask, 'ORIGINAL_UPB'] / (real.loc[mask, 'LTV'] / 100)
real.loc[mask, 'ESTIMATED_LTV'] = (real.loc[mask, 'CURRENT_ACTUAL_UPB'] / valeur_bien) * 100

# Nettoyage
real = real.drop(columns=['ORIGINAL_UPB', 'LTV'])

# Vérification
real['ESTIMATED_LTV'].isna().sum()

np.int64(2200)

In [50]:
mask = real['ESTIMATED_LTV'].isna()
print(real[mask][['LOAN_SEQUENCE_NUMBER', 'CURRENT_ACTUAL_UPB']].head(10))
print(f"\nLOAN_SEQUENCE_NUMBER uniques : {real[mask]['LOAN_SEQUENCE_NUMBER'].nunique()}")

        LOAN_SEQUENCE_NUMBER  CURRENT_ACTUAL_UPB
1708956         F07Q10037840           125000.00
1708957         F07Q10037840           125000.00
1708958         F07Q10037840           125000.00
1708959         F07Q10037840           125000.00
1708960         F07Q10037840           125000.00
1708961         F07Q10037840           124000.00
1708962         F07Q10037840           124000.00
1708963         F07Q10037840           124215.10
1708964         F07Q10037840           124100.48
1708965         F07Q10037840           123985.23

LOAN_SEQUENCE_NUMBER uniques : 38


In [51]:
# Combien de prêts uniques ?
print(real[mask]['LOAN_SEQUENCE_NUMBER'].nunique())

# Sont-ils dans l'origination ?
orphelins = real[mask]['LOAN_SEQUENCE_NUMBER'].unique()
print(f"Présents dans origination : {orig['LOAN_SEQUENCE_NUMBER'].isin(orphelins).sum()}")

38
Présents dans origination : 38


In [52]:
# Vérifier ce qui manque dans l'origination pour ces prêts
orphelins = real[real['ESTIMATED_LTV'].isna()]['LOAN_SEQUENCE_NUMBER'].unique()
print(orig[orig['LOAN_SEQUENCE_NUMBER'].isin(orphelins)][['LOAN_SEQUENCE_NUMBER', 'ORIGINAL_UPB', 'LTV']].head(10))

       LOAN_SEQUENCE_NUMBER  ORIGINAL_UPB  LTV
26460          F07Q10037840        125000  NaN
34769          F07Q10049414         16000  NaN
38812          F07Q10054647         40000  NaN
46866          F07Q10066891         60000  NaN
61304          F07Q10088035         94000  NaN
75994          F07Q10110000         73000  NaN
84552          F07Q10120533         55000  NaN
102027         F07Q10143798        105000  NaN
102038         F07Q10143809         28000  NaN
102449         F07Q10144235         71000  NaN


                ORIGINAL_UPB est renseigné mais LTV est NaN — on ne peut pas calculer la valeur du bien sans LTV.
                troi options :

                    Imputer LTV par la médiane du segment avant de recalculer — mais c'est une double approximation
                    Imputer ESTIMATED_LTV directement par la médiane du segment sur les prêts similaires
                    Suppresion des pret a l'origination et en performance

                pour l'option 3 — moins de chaîne d'approximations. C'est seulement 38 prêts sur 1.39 million, l'impact est négligeable.

In [53]:
# 1. Identifier les prêts avec ESTIMATED_LTV manquant
mask_ltv_nan = real['ESTIMATED_LTV'].isna()

# 2. Extraire les séquences concernées
loans_to_drop = real[mask_ltv_nan]['LOAN_SEQUENCE_NUMBER'].unique()
print(f"Prêts à supprimer : {len(loans_to_drop)}")
print(f"Lignes concernées : {mask_ltv_nan.sum()}")

# 3. Suppression dans l'origination
orig = orig[~orig['LOAN_SEQUENCE_NUMBER'].isin(loans_to_drop)]
print(f"Origination — Prêts restants : {orig['LOAN_SEQUENCE_NUMBER'].nunique():,}")

# 4. Suppression dans la performance
real = real[~real['LOAN_SEQUENCE_NUMBER'].isin(loans_to_drop)]
print(f"Performance — Prêts restants : {real['LOAN_SEQUENCE_NUMBER'].nunique():,}")

# 5. Vérification
print(f"ESTIMATED_LTV NaN : {real['ESTIMATED_LTV'].isna().sum()}")

Prêts à supprimer : 38
Lignes concernées : 2200
Origination — Prêts restants : 1,390,970
Performance — Prêts restants : 1,390,970
ESTIMATED_LTV NaN : 0


In [54]:
# Identifier les prêts orphelins restants
orphelins_perf = real[real['ESTIMATED_LTV'].isna()]['LOAN_SEQUENCE_NUMBER'].unique()

# Sont-ils dans l'origination ?
print(f"Présents dans origination : {orig['LOAN_SEQUENCE_NUMBER'].isin(orphelins_perf).sum()}")
print(f"Prêts uniques concernés : {len(orphelins_perf)}")

Présents dans origination : 0
Prêts uniques concernés : 0


MODIFICATION_FLAG

            indique si le prêt a fait l'objet d'une modification contractuelle
             changement de taux, allongement de durée, réduction de capital — dans le cadre d'un programme de rétention.
                C'est un signal de fragilité — un prêt modifié signifie que le client n'a pas pu honorer ses engagements initiaux sans aide.

            97.5% de NaN — mais ce n'est pas un problème, c'est sémantiquement correct.


In [55]:
real['MODIFICATION_FLAG'] = real['MODIFICATION_FLAG'].fillna('N')

ZERO_BALANCE_EFFECTIVE_DATE → 98.5% — NaN = prêts actifs, garder en NaT

DUE_DATE_OF_LAST_PAID_INSTALLMENT → 99.3% — NaN = client à jour, garder en NaT

INTEREST_RATE_STEP_INDICATOR

                Elle indique si le taux d'intérêt du prêt est en phase de transition — c'est-à-dire si le taux va augmenter ou diminuer au
                prochain palier contractuel.
                Elle concerne uniquement les ARM (taux variable) qui ont des paliers de taux programmés dans le contrat.
                100% de NaN sur notre dataset — notre portefeuille est majoritairement composé de FRM.

                Resolution suppression complete

In [56]:
# Usage d'un prefixw pour recupere l'année de l'origination
real['VINTAGE_TEMP'] = real['LOAN_SEQUENCE_NUMBER'].str[1:3]

print(real.groupby('VINTAGE_TEMP')['INTEREST_RATE_STEP_INDICATOR'].apply(lambda x: x.isna().mean()).round(3))
real = real.drop(columns=['VINTAGE_TEMP'])

VINTAGE_TEMP
07    0.934
10    0.990
20    0.996
25    1.000
Name: INTEREST_RATE_STEP_INDICATOR, dtype: float64


In [57]:
real['INTEREST_RATE_STEP_INDICATOR'] = real['INTEREST_RATE_STEP_INDICATOR'].fillna('N')

In [58]:
real[ELIGIBLE_COLUMNS].isna().mean().round(6)

LOAN_SEQUENCE_NUMBER                  0.000000
MONTHLY_REPORTING_PERIOD              0.000000
CURRENT_ACTUAL_UPB                    0.000000
CURRENT_LOAN_DELINQUENCY_STATUS       0.000000
LOAN_AGE                              0.000000
REMAINING_MONTHS_TO_LEGAL_MATURITY    0.000000
MODIFICATION_FLAG                     0.000000
ZERO_BALANCE_CODE                     0.000000
ZERO_BALANCE_EFFECTIVE_DATE           0.985597
CURRENT_INTEREST_RATE                 0.000000
CURRENT_NON_INTEREST_BEARING_UPB      0.000000
DUE_DATE_OF_LAST_PAID_INSTALLMENT     0.993739
INTEREST_RATE_STEP_INDICATOR          0.000000
ESTIMATED_LTV                         0.000000
DELINQUENCY_DUE_TO_DISASTER           0.000000
BORROWER_ASSISTANCE_STATUS_CODE       0.000000
INTEREST_BEARING_UPB                  0.000000
VINTAGE                               0.000000
LOAN_SEQUENCE_NUMBER                  0.000000
dtype: float64

Save Data

In [59]:
# Supprimer les colonnes dupliquées
real = real.loc[:, ~real.columns.duplicated()]



In [60]:
real.columns.tolist()

['LOAN_SEQUENCE_NUMBER',
 'MONTHLY_REPORTING_PERIOD',
 'CURRENT_ACTUAL_UPB',
 'CURRENT_LOAN_DELINQUENCY_STATUS',
 'LOAN_AGE',
 'REMAINING_MONTHS_TO_LEGAL_MATURITY',
 'DEFECT_SETTLEMENT_DATE',
 'MODIFICATION_FLAG',
 'ZERO_BALANCE_CODE',
 'ZERO_BALANCE_EFFECTIVE_DATE',
 'CURRENT_INTEREST_RATE',
 'CURRENT_NON_INTEREST_BEARING_UPB',
 'DUE_DATE_OF_LAST_PAID_INSTALLMENT',
 'MI_RECOVERIES',
 'NET_SALE_PROCEEDS',
 'NON_MI_RECOVERIES',
 'TOTAL_EXPENSES',
 'LEGAL_COSTS',
 'MAINTENANCE_AND_PRESERVATION_COSTS',
 'TAXES_AND_INSURANCE',
 'MISCELLANEOUS_EXPENSES',
 'ACTUAL_LOSS_CALCULATION',
 'CUMULATIVE_MODIFICATION_COST',
 'INTEREST_RATE_STEP_INDICATOR',
 'PAYMENT_DEFERRAL_FLAG',
 'ESTIMATED_LTV',
 'ZERO_BALANCE_REMOVAL_UPB',
 'DELINQUENT_ACCRUED_INTEREST',
 'DELINQUENCY_DUE_TO_DISASTER',
 'BORROWER_ASSISTANCE_STATUS_CODE',
 'CURRENT_MONTH_MODIFICATION_COST',
 'INTEREST_BEARING_UPB',
 'VINTAGE']

In [61]:
real[to_drop].to_csv('/Users/macbookpro/platform/Backend/data/working/svcg_NON_ELLIGIBLE.csv',sep=',')

In [63]:
def harmonize_dtypes(df):
    for col in df.columns:
        # Colonnes object (string mixte)
        if df[col].dtype == object:
            df[col] = df[col].astype(str).replace('nan', None)
        # Colonnes numériques → garder telles quelles
        # Colonnes datetime → garder telles quelles
    return df


df = harmonize_dtypes(real.copy())
df.to_parquet('/Users/macbookpro/platform/Backend/data/processed/svcg_ALL_COLUMNS_CLEANED.parquet', index=False)
print("Origination sauvegardée ")

Origination sauvegardée 


In [ ]:
/Users/macbookpro/platform/Backend/data/processed/svcg_All_columns.parquet